### PR TITLE
merge_asof and prefix_reduction

### PR DESCRIPTION
**Pull request by codercody**
- [ ] Tests added / passed
- [ ] Passes `flake8 dask`
- [ ] Added a parallel prefix sum algorithm to core
- [ ] Fixed a small bug with methods.boundary_slice (they forgot one edge case)
- [ ] Added merge_asof

For merge_asof(left, right), the algorithm is to:
1. slice up left so that its divisions are left.divisions U right.divisions
2. pair up the ith part on the left to the jth part on the right when the former is a subset (in terms of indices) of the latter
3. pad the right part with the last row of partition j-1 (roughly) and the first row of partition j+1 (roughly) in case the data needs to be retrieved from another partition
4. merge all the paired padded parts and combine

merge_asof_demo

The things I know I messed up in this PR (and would like help fixing) are
- I didn't modify the sphinx code because I didn't understand how to
- My test cases are in a completely separate file from test_multi.py
- I didn't include any test cases for the prefix sum yet since I didn't expect to turn it into a method of its own until the very end I decided to (though I did thoroughly test it beforehand)

### CODE REVIEW

**Comment by jakirkham** [comment]
@jcrist, would you be able to take a look at this?

**Comment by mrocklin** [comment]
Hi @codercody , this looks like exciting work.  I have a few rough suggestions, though I suspect that folks like @TomAugspurger or @jrbourbeau (who know a bit more about dask dataframe) may have more.

1.  If possible, it would be good not to include moderately large testing files in the repository.  This makes things a little bloated.  Ideally you would be able to quickly generate some artificial data as part of a test. 
2.  Similarly, those tests should probably be pretty fast and easy to understand, so smaller testing datasets would be welcome.
3.  I suspect that thee linter will yell at you.  You might want to run `flake8 dask` on your code in order to get a sense of expectations.  This is also available here: https://travis-ci.org/dask/dask/jobs/541084621

File: dask/dataframe/tests/test_multi.py [hunk 1]
```diff
@@ -262,6 +264,168 @@ def test_sequential_joins():
...
+    a = dd.from_pandas(A, npartitions=2)
+    B = pd.DataFrame({'a': [1, 2, 3, 6, 7], 'right_val': [1, 2, 3, 6, 7]})
+    b = dd.from_pandas(B, npartitions=2)
+
+    C = pd.merge_asof(A, B, on='a').set_index('a')
+    c = merge_asof(a, b, on='a')
+    assert_eq(c, C)
+
+
+def test_merge_asof_on_no_exact_matches():
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py]: This could be parametrized and shared with the previous test I think. Something like

````python
@pytest.mark.parametrize('allow_exact_matches', [True, False])
def test_merge_asof_on(allow_exact_matches):
    ...
```

File: dask/dataframe/tests/test_multi.py [hunk 2]
```diff
@@ -262,6 +264,168 @@ def test_sequential_joins():
...
+    a = dd.from_pandas(A, npartitions=2)
+    B = pd.DataFrame({'a': [1, 2, 3, 6, 7], 'right_val': [1, 2, 3, 6, 7]})
+    b = dd.from_pandas(B, npartitions=2)
+
+    C = pd.merge_asof(A, B, on='a', direction='forward').set_index('a')
+    c = merge_asof(a, b, on='a', direction='forward')
+    assert_eq(c, C)
+
+
+def test_merge_asof_on_nearest():
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py]: Same comment about parametrizing.

File: dask/dataframe/tests/test_multi.py [hunk 3]
```diff
@@ -262,6 +264,168 @@ def test_sequential_joins():
...
+    assert_eq(c, C)
+
+
+def test_merge_asof_on_no_exact_matches():
+    A = pd.DataFrame({'a': [1, 5, 10], 'left_val': ['a', 'b', 'c']})
+    a = dd.from_pandas(A, npartitions=2)
+    B = pd.DataFrame({'a': [1, 2, 3, 6, 7], 'right_val': [1, 2, 3, 6, 7]})
+    b = dd.from_pandas(B, npartitions=2)
+
+    C = pd.merge_asof(A, B, on='a', allow_exact_matches=False).set_index('a')
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py]: Why is the `set-index` needed here? I haven't looked at the implementation yet, forgive me if that's explained there.

File: dask/dataframe/tests/test_multi.py [hunk 4]
```diff
@@ -262,6 +264,168 @@ def test_sequential_joins():
...
+    B = pd.DataFrame({'time': times_B, 'ticker': tickers_B, 'price': prices_B, 'quantity': quantities_B})
+    b = dd.from_pandas(B, npartitions=3)
+
+    C = pd.merge_asof(B, A, on='time', by='ticker').set_index('time')
+    c = merge_asof(b, a, on='time', by='ticker')
+    assert_eq(c, C)
+
+
+def test_merge_asof_on_by_tolerance():
+    times_A = [parse(d) for d in ['2016-05-25 13:30:00.023',
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py]: Use `pd.to_datetime(...)` on the list, rather than `parse(d)` on each element. Here and in the next test.

File: dask/dataframe/multi.py [hunk 5]
```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
...
+    return new_dd_object(graph, name, meta, left.divisions)
+
+
+@wraps(pd.merge_asof)
+def merge_asof(left, right, on=None, left_on=None, right_on=None,
+               left_index=False, right_index=False, by=None, left_by=None,
+               right_by=None, suffixes=('_x', '_y'), tolerance=None,
+               allow_exact_matches=True, direction='backward'):
+    if direction not in ['backward', 'forward', 'nearest']:
+        raise ValueError("Invalid merge_asof direction. Choose from 'backward',\
```
Comment (TomAugspurger) [inline dask/dataframe/multi.py]: This will have weird spacing in the message. Something like

```
        raise ValueError("Invalid merge_asof direction. Choose from 'backward'"
                         " 'forward', or 'nearest'")


```

instead.

File: dask/dataframe/multi.py [hunk 6]
```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
...
+              'left_index': left_index, 'right_index': right_index,
+              'by': by, 'left_by': left_by, 'right_by': right_by,
+              'suffixes': suffixes, 'tolerance': tolerance,
+              'allow_exact_matches': allow_exact_matches,
+              'direction': direction}
+
+    if left is None or right is None:
+        raise ValueError("Cannot merge_asof on empty DataFrames")
+
+    if isinstance(left, pd.DataFrame) and isinstance(right, pd.DataFrame):
```
Comment (TomAugspurger) [inline dask/dataframe/multi.py]: @mrocklin what's best for cuDF here? Would a dispatchable `merge_asof`, and replace these checks with `is_dataframe_like` be best?

File: dask/dataframe/multi.py [hunk 7]
```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
...
+        left_on = right_on = on
+    for o in [left_on, right_on]:
+        if isinstance(o, _Frame):
+            raise NotImplementedError(
+                "Dask collections not currently allowed in merge columns")
+
+    if not is_dask_collection(left):
+        left = from_pandas(left, npartitions=1)
+    if left_on is not None:
+        left = left.set_index(left_on, sorted=True)
```
Comment (TomAugspurger) [inline dask/dataframe/multi.py]: `left` and `right` are coming straight from the user, right? Saying `sorted=True` when it's not in fact sorted is dangerous. I'm not sure we should do that here.

Comment (codercody) [reply dask/dataframe/multi.py]: It sure is dangerous. However, in the documentation for merge_asof it says the input data must be sorted. I don't see any good way of verifying that it's sorted before doing this operation without messing up the complexity, so I would just be inclined to trust the user. https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.merge_asof.html

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (codercody) [reply dask/dataframe/tests/test_multi.py]: Sorry. When Pandas does a merge_asof on a particular column, the resulting dataframe gets reindexed 0, 1, .... This is the one place where my implementation differs from Pandas': if your merge_asof on a particular column using mine, it sets the index as that on column. I could alternatively do a reset_index() in my implementation but I find it kind of silly.

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -4939,3 +4939,137 @@ def meta_warning(df):
...
+
+    If df has partitions [P1, P2, ..., Pn], then returns the DataFrame with
+    partitions [f(z, P1), f(f(z, P1), P2), f(f(f(z, P1), P2), P3), ...]
+
+    If you don't specify identity, make sure f handles when either argument is
+    None!
+
+    Parameters
+    ----------
+    f : 'a x 'a -> 'a
```
[the remark this answers is not in the document]
Comment (codercody) [reply dask/dataframe/core.py]: Is there any appropriate place for me to specify the typing so that it's clear to users? Or should I just leave it out and have them figure it out themselves?

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (codercody) [reply dask/dataframe/tests/test_multi.py]: Furthermore, this is causing Travis CI to mess up, since I guess this causes some index to be permuted in a different way.

```
E       AssertionError: Index are different
E       
E       Index values are different (100.0 %)
E       [left]:  Index([u'price', u'quantity', u'ticker', u'ask', u'bid'], dtype='object')
E       [right]: Index([u'ask', u'bid', u'price', u'quantity', u'ticker'], dtype='object')
```

Do you know what I should do to fix this to pass Travis CI?

File: dask/dataframe/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply dask/dataframe/core.py]: In the description is fine.

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (TomAugspurger) [reply dask/dataframe/tests/test_multi.py]: Ideally we would match pandas output.

We may not be able to entirely in this case. This might be a case where we need each partition to have its own `RangeIndex(0, len(partition))`.

File: dask/dataframe/multi.py (hunk 7 above)
Comment (TomAugspurger) [reply dask/dataframe/multi.py]: Pandas validates the sorting though, 

```python
In [7]: >>> pd.merge_asof(left, right[::-1], on='a')
   ...:
   ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-481523b2ecc3> in <module>
----> 1 pd.merge_asof(left, right[::-1], on='a')
      2

~/sandbox/pandas/pandas/core/reshape/merge.py in merge_asof(left, right, on, left_on, right_on, left_index, right_index, by, left_by, right_by, suffixes, tolerance, allow_exact_matches, direction)
    464                     allow_exact_matches=allow_exact_matches,
    465                     direction=direction)
--> 466     return op.get_result()
    467
    468

~/sandbox/pandas/pandas/core/reshape/merge.py in get_result(self)
   1295
   1296     def get_result(self):
-> 1297         join_index, left_indexer, right_indexer = self._get_join_info()
   1298
   1299         # this is a bit kludgy

~/sandbox/pandas/pandas/core/reshape/merge.py in _get_join_info(self)
    758         else:
    759             (left_indexer,
--> 760              right_indexer) = self._get_join_indexers()
    761
    762             if self.right_index:

~/sandbox/pandas/pandas/core/reshape/merge.py in _get_join_indexers(self)
   1535                 raise ValueError(msg_missings.format(side='right'))
   1536             else:
-> 1537                 raise ValueError(msg_sorted.format(side='right'))
   1538
   1539         # initial type conversion as needed

ValueError: right keys must be sorted
```

Can you add a test ensuring that `dd.merge_asof` raises in a similar case for dask arrays?

File: dask/dataframe/multi.py (hunk 6 above)
Comment (mrocklin) [reply dask/dataframe/multi.py]: I think that we should use `is_dataframe_like` universally.

I think that we should continue to use `pd.merge_asof` until some other pandas implementation supports the operation, at that point I think that we would either try to add a method, or a dispatchable function.  I wouldn't worry about this until that happens though.

File: dask/dataframe/multi.py (hunk 7 above)
Comment (mrocklin) [reply dask/dataframe/multi.py]: I'm not sure we can reasonably verify this without doing a full-pass computation.

File: dask/dataframe/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrocklin) [reply dask/dataframe/core.py]: To be clear, you should do something like this:

```
f : callable
    An associative function of the form ...
```

File: dask/dataframe/multi.py [hunk 9]
```diff
@@ -419,6 +420,233 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
...
+    def none_to_empty(df):
+        return empty if df is None else df
+
+    if by is None:
+        def last_tail(L, R):
+            if R is None or R.empty:
+                return L
+            return R.tail(1)
+
+        return prefix_reduction(last_tail, ddf, postprocess=none_to_empty)
```
Comment (mrocklin) [inline dask/dataframe/multi.py]: Passing dynamically generated functions into dask graphs is a bit suboptimal, particularly when using a distributed cluster.  It's significantly more expensive to serialize a dynamically generated function in Python than one that is defined within a module.

Comment (mrocklin) [reply dask/dataframe/multi.py]: To be more clear, can I ask you to not define functions within other functions, if they are going to be passed into a task graph.  Functions like `last_tail` or `none_to_empty` defined just above should be moved outside of the `compute_tails` definition if they are going to end up in the task graph (which I think that they are).  This happens in a few other places as well.

What you've done works, but is slightly less efficient for reasons that are probably not so important to get into.

File: dask/dataframe/multi.py (hunk 6 above)
Comment (mrocklin) [reply dask/dataframe/multi.py]: So, to be clear @codercody , if you can replace checks like 

```python
isinstance(df, pd.DataFrame)
```

with 

```python
is_dataframe_like(df)
```

That would be welcome.

File: dask/dataframe/multi.py [hunk 10]
```diff
@@ -419,6 +420,233 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
                          indicator=indicator)
 
 
+###############################################################
+# ASOF Join
+###############################################################
```
Comment (mrocklin) [inline dask/dataframe/multi.py]: Given the amount of logic here, I think it would be fine to move this to a separate file, if you wanted (either way is fine).

File: dask/dataframe/multi.py (hunk 7 above)
Comment (mrocklin) [reply dask/dataframe/multi.py]: @TomAugspurger what do you think we should do here?  I suspect that verifying sortedness is hard to do lazily.

Comment (TomAugspurger) [reply dask/dataframe/multi.py]: Since pandas raises, I think we’re ok. Would be good to add a test to verify.

Comment (TomAugspurger) [reply dask/dataframe/multi.py]: To be clear, pandas would be the one to raise when we execute the graph.

File: dask/dataframe/multi.py (hunk 9 above)
Comment (codercody) [reply dask/dataframe/multi.py]: I’m on it (but I’m away from my laptop until Monday, sorry).

File: dask/dataframe/multi.py (hunk 7 above)
Comment (codercody) [reply dask/dataframe/multi.py]: Pandas wouldn't raise if the out-of-order values cross the partitions. If we _must_ check this then we could always `map_partitions` `min` and `max` of the `by` column to each partition and check the divisions. But I would rather leave the burden on the user to ensure that their data is ordered then mess up our runtime.

File: dask/dataframe/multi.py (hunk 6 above)
Comment (codercody) [reply dask/dataframe/multi.py]: I was trying to copy line 346 in the `merge` implementation, where they separately handled when `left` and `right` are actually `pd.DataFrame`s and not `dd.DataFrame`s. If you don't want me to handle this case, then I'm just going to delete this check altogether. Is that fine?

File: dask/dataframe/multi.py (hunk 10 above)
Comment (codercody) [reply dask/dataframe/multi.py]: I prefer it in the `multi.py` file. It's only ~200 lines just like `merge` and `concat` are.

**Comment by mrocklin** [comment]
@codercody I see that you reacted with the Confused emoji to @TomAugspurger 's recent comment below:

> It's possible you're relying on dict ordering here, while we support Python 3.5.

To expand on this, it looks like the tests are failing, but only in Python 3.5, and not in any newer Python versions.  You can see the failure here: https://travis-ci.org/dask/dask/jobs/544385948

<details>

```
____________________________ test_merge_asof_on_by _____________________________
[gw2] linux -- Python 3.5.6 /home/travis/miniconda/envs/test-environment/bin/python
    def test_merge_asof_on_by():
        times_A = [pd.to_datetime(d) for d in ['2016-05-25 13:30:00.023',
                                               '2016-05-25 13:30:00.023',
                                               '2016-05-25 13:30:00.030',
                                               '2016-05-25 13:30:00.041',
                                               '2016-05-25 13:30:00.048',
                                               '2016-05-25 13:30:00.049',
                                               '2016-05-25 13:30:00.072',
                                               '2016-05-25 13:30:00.075']]
        tickers_A = ['GOOG', 'MSFT', 'MSFT', 'MSFT', 'GOOG', 'AAPL', 'GOOG', 'MSFT']
        bids_A = [720.50, 51.95, 51.97, 51.99, 720.50, 97.99, 720.50, 52.01]
        asks_A = [720.93, 51.96, 51.98, 52.00, 720.93, 98.01, 720.88, 52.03]
        times_B = [pd.to_datetime(d) for d in ['2016-05-25 13:30:00.023',
                                               '2016-05-25 13:30:00.038',
                                               '2016-05-25 13:30:00.048',
                                               '2016-05-25 13:30:00.048',
                                               '2016-05-25 13:30:00.048']]
        tickers_B = ['MSFT', 'MSFT', 'GOOG', 'GOOG', 'AAPL']
        prices_B = [51.95, 51.95, 720.77, 720.92, 98.00]
        quantities_B = [75, 155, 100, 100, 100]
    
        A = pd.DataFrame({'time': times_A, 'ticker': tickers_A, 'bid': bids_A, 'ask': asks_A})
        a = dd.from_pandas(A, npartitions=4)
        B = pd.DataFrame({'time': times_B, 'ticker': tickers_B, 'price': prices_B, 'quantity': quantities_B})
        b = dd.from_pandas(B, npartitions=3)
    
        C = pd.merge_asof(B, A, on='time', by='ticker').set_index('time')
        c = merge_asof(b, a, on='time', by='ticker')
>       assert_eq(c, C)
dask/dataframe/tests/test_multi.py:342: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/dataframe/utils.py:697: in assert_eq
    a = _check_dask(a, check_names=check_names, check_dtypes=check_dtypes)
dask/dataframe/utils.py:653: in _check_dask
    tm.assert_index_equal(dsk.columns, result.columns)
../../../miniconda/envs/test-environment/lib/python3.5/site-packages/pandas/util/testing.py:922: in assert_index_equal
    raise_assert_detail(obj, msg, left, right)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
obj = 'Index', message = 'Index values are different (100.0 %)'
left = Index(['price', 'quantity', 'ticker', 'ask', 'bid'], dtype='object')
right = Index(['ask', 'bid', 'price', 'quantity', 'ticker'], dtype='object')
diff = None
    def raise_assert_detail(obj, message, left, right, diff=None):
        if isinstance(left, np.ndarray):
            left = pprint_thing(left)
        elif is_categorical_dtype(left):
            left = repr(left)
        if isinstance(right, np.ndarray):
            right = pprint_thing(right)
        elif is_categorical_dtype(right):
            right = repr(right)
    
        msg = """{obj} are different
    
    {message}
    [left]:  {left}
    [right]: {right}""".format(obj=obj, message=message, left=left, right=right)
    
        if diff is not None:
            msg += "\n[diff]: {diff}".format(diff=diff)
    
>       raise AssertionError(msg)
E       AssertionError: Index are different
E       
E       Index values are different (100.0 %)
E       [left]:  Index(['price', 'quantity', 'ticker', 'ask', 'bid'], dtype='object')
E       [right]: Index(['ask', 'bid', 'price', 'quantity', 'ticker'], dtype='object')
```

</details>

Given the fact that this is an ordering thing, and that it's only on Python 3.5, Tom was guessing that this was due to expecting that dicts are consistently ordered, which is something that changed between 3.5 and 3.6.  In Python 3.6 calling `list(my_dict)` always gives the same result, and is consistent with the order of the input. 

To make things concrete, if you create an environment with Python 3.5 do your tests pass for you?

Also, as a quick note, many of the maintainers track PRs by e-mail.  We get notifications when someone pushes a new commit or makes a comment, but not when someone reacts to a comment with an emoji, hence the delay in response here.

**Comment by mrocklin** [comment]
@TomAugspurger other than the dict ordering issue is this ready to go in, or does it still need more work?

This is one of the few substantial outstanding PRs that we have before we might use black.  I have a slight preference to black-ify things before the release, which I would ideally like to do soon.  (though that's somewhat arbitrary, black could slip, or we could redo this PR after black.

File: dask/dataframe/tests/test_multi.py [hunk 11]
```diff
@@ -262,6 +262,150 @@ def test_sequential_joins():
...
+    assert_eq(c, C)
+
+
+def test_merge_asof_on():
+    A = pd.DataFrame({'a': [1, 5, 10], 'left_val': ['a', 'b', 'c']})
+    a = dd.from_pandas(A, npartitions=2)
+    B = pd.DataFrame({'a': [1, 2, 3, 6, 7], 'right_val': [1, 2, 3, 6, 7]})
+    b = dd.from_pandas(B, npartitions=2)
+
+    C = pd.merge_asof(A, B, on='a').set_index('a')
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py]: This still concerns me. I'm looking into why it's necessary now.

**Comment by TomAugspurger** [comment]
@codercody apologies for the flurry of commits. Can yo take a look at https://github.com/dask/dask/pull/4877/commits/0e60df1ee21a8383c458c669f28b684320891feb and confirm that it looks OK?

That matches the output, other than the index. With Dask's merge_asof, each partition has its on `RangeIndex(0, n_rows)`. We need to document the difference in behavior somewhere, but the `wraps` we're using to copy the docstring doesn't allow that. The usual `utils.derived_from` only works with classes at the moment.

Docs probably aren't a blocker.

**Comment by TomAugspurger** [comment]
Sorry, I seem to have broken things (right now I think it's an issue with `concat` no longer implicitly sorting. I'll take a look in a couple hours).

**Comment by TomAugspurger** [comment]
https://github.com/dask/dask/pull/4877/commits/fd6cef36b15dced5effdf96dc7c11e89e6a357b8 should pass. This is probably good to go, though I would like to see the docs issue addressed eventually.

**Comment by TomAugspurger** [comment]
Still seem to have a column order issue with pandas 0.21.x.

**Comment by mrocklin** [comment]
Green check mark :)

**Comment by TomAugspurger** [comment]
Alrighty, here goes.

Thanks @codercody, nice work!
